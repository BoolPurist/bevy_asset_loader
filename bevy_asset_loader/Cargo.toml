[package]
name = "bevy_asset_loader"
version = "0.8.0"
authors = ["Niklas Eicker <git@nikl.me>"]
edition = "2021"
license = "MIT OR Apache-2.0"
description = "Bevy plugin for asset loading"
repository = "https://github.com/NiklasEi/bevy_asset_loader"
homepage = "https://github.com/NiklasEi/bevy_asset_loader"
documentation = "https://docs.rs/bevy_asset_loader"
keywords = ["bevy", "gamedev", "asset", "assets"]
categories = ["game-development"]
readme = "../README.md"

[features]
# This feature requires bevy's TextureAtlas and StandardMaterial assets to be available ("bevy/render")
render = ["bevy_asset_loader_derive/render"]

[dependencies]
bevy = { version = "0.6", default-features = false }
bevy_asset_loader_derive = { version = "=0.8.0", path = "../bevy_asset_loader_derive" }

[dev-dependencies]
bevy = { version = "0.6", features = ["vorbis"] }
trybuild = "1.0"

[[example]]
name = "two_collections"
path = "examples/two_collections.rs"

[[example]]
name = "atlas_from_grid"
path = "examples/atlas_from_grid.rs"
required-features = ["render"]

[[example]]
name = "init_resource"
path = "examples/init_resource.rs"

[[example]]
name = "dynamic_asset"
path = "examples/dynamic_asset.rs"

[[example]]
name = "color_material"
path = "examples/standard_material.rs"
required-features = ["render"]
