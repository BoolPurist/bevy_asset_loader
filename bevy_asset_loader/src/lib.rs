//! The goal of this crate is to offer an easy way for bevy games to load all their assets.
//!
//! ```edition2018
//! # use bevy_asset_loader::{AssetLoaderPlugin, AssetCollection};
//! # use bevy_kira_audio::{AudioPlugin, AudioSource, Audio};
//! # use bevy::prelude::*;
//! fn main() {
//!     App::build()
//!         .add_state(MyStates::Load)
//!         .add_plugins(DefaultPlugins)
//!         .add_plugin(AudioPlugin)
//!         .add_plugin(AssetLoaderPlugin::<MyAudioAssets, _>::new(
//!             MyStates::Load,
//!             MyStates::Next,
//!         ))
//!         .add_system_set(SystemSet::on_update(MyStates::Next).with_system(use_asset_handles.system()))
//!         .run();
//! }
//!
//! #[derive(AssetCollection)]
//! struct MyAudioAssets {
//!     #[asset(path = "walking.ogg")]
//!     walking: Handle<AudioSource>,
//!     #[asset(path = "flying.ogg")]
//!     flying: Handle<AudioSource>
//! }
//!
//! // since this function runs in [MyState::Next], we know our assets are
//! // loaded and [MyAudioAssets] is a resource
//! fn use_asset_handles(audio_assets: Res<MyAudioAssets>, audio: Res<Audio>) {
//!     audio.play(audio_assets.flying.clone());
//! }
//!
//! #[derive(Clone, Eq, PartialEq, Debug, Hash)]
//! enum MyStates {
//!     Load,
//!     Next
//! }
//! ```

#![forbid(unsafe_code)]
#![warn(unused_imports, missing_docs)]

pub use bevy_asset_loader_derive::AssetCollection;

use bevy::app::{AppBuilder, Plugin};
use bevy::asset::{AssetServer, HandleId, HandleUntyped, LoadState};
use bevy::ecs::component::Component;
use bevy::ecs::schedule::State;
use bevy::ecs::system::IntoSystem;
use bevy::prelude::{Commands, Res, ResMut, SystemSet};
use std::fmt::Debug;
use std::hash::Hash;
use std::marker::PhantomData;

/// A Bevy plugin to load asset collections
///
/// Configure it with a struct implementing [AssetCollection] and two [bevy::ecs::schedule::State]s
/// ```edition2018
/// # use bevy_asset_loader::{AssetLoaderPlugin, AssetCollection};
/// # use bevy::prelude::*;
/// fn main() {
///    let mut app = App::build();
///    app
///         .add_plugins(DefaultPlugins)
///         .add_plugin(AssetLoaderPlugin::<MyAssets, _>::new(
///             MyStates::Load,
///             MyStates::Next,
///         ));
///    app.run();
/// }
///
/// # #[derive(AssetCollection)]
/// # struct MyAssets {
/// #     #[asset(path = "player.png")]
/// #     player: Handle<Texture>,
/// #     #[asset(path = "tree.png")]
/// #     tree: Handle<Texture>
/// # }
/// #
/// # #[derive(Clone, Eq, PartialEq, Debug, Hash)]
/// # enum MyStates {
/// #     Load,
/// #     Next
/// # }
/// ```
pub struct AssetLoaderPlugin<A, T> {
    /// The state during which the assets should be loaded
    on: T,
    /// The state the plugin will set when the assets are loaded
    next: T,
    marker: PhantomData<A>,
}

impl<A, T> AssetLoaderPlugin<A, T>
where
    A: AssetCollection,
    T: Component + Debug + Clone + Eq + Hash,
{
    /// Create a new instance of [AssetLoaderPlugin]
    ///
    /// ```edition2018
    /// # use bevy_asset_loader::{AssetLoaderPlugin, AssetCollection};
    /// # use bevy::prelude::*;
    /// fn main() {
    ///    let mut app = App::build();
    ///    app
    ///         .add_plugins(DefaultPlugins)
    ///         .add_plugin(AssetLoaderPlugin::<MyAssets, _>::new(
    ///             MyStates::Load,
    ///             MyStates::Next,
    ///         ));
    ///    app.run();
    /// }
    ///
    /// # #[derive(AssetCollection)]
    /// # struct MyAssets {
    /// #     #[asset(path = "player.png")]
    /// #     player: Handle<Texture>,
    /// #     #[asset(path = "tree.png")]
    /// #     tree: Handle<Texture>
    /// # }
    /// #
    /// # #[derive(Clone, Eq, PartialEq, Debug, Hash)]
    /// # enum MyStates {
    /// #     Load,
    /// #     Next
    /// # }
    /// ```
    pub fn new(on: T, next: T) -> AssetLoaderPlugin<A, T> {
        Self {
            on,
            next,
            marker: PhantomData::<A>,
        }
    }
}

/// Trait to mark a struct as a collection os assets
///
/// Derive is supported for structs with named fields of the type [bevy::asset::Handle]<T: [bevy::asset::Asset]>
/// Each field needs to be annotated with ``#[asset(path = "path/to/asset.png")]``
/// ```edition2018
/// # use bevy::asset::Handle;
/// #[derive(AssetCollection)]
/// struct MyAssets {
///     #[asset(path = "player.png")]
///     player: Handle<Texture>,
///     #[asset(path = "tree.png")]
///     tree: Handle<Texture>
/// }
/// ```
pub trait AssetCollection: Component {
    /// Create a new AssetCollection from the [bevy::asset::AssetServer]
    fn create(asset_server: &Res<AssetServer>) -> Self;
    /// Start loading all the assets in the collection
    fn load(asset_server: &Res<AssetServer>) -> Vec<HandleUntyped>;
}

struct LoadingAssetHandles<A: Component> {
    handles: Vec<HandleId>,
    marker: PhantomData<A>,
}

struct AssetLoaderNextState<T> {
    next: T,
}

impl<Assets, State> Plugin for AssetLoaderPlugin<Assets, State>
where
    Assets: AssetCollection,
    State: Component + Debug + Clone + Eq + Hash,
{
    fn build(&self, app: &mut AppBuilder) {
        app.insert_resource(AssetLoaderNextState::<State> {
            next: self.next.clone(),
        })
        .add_system_set(
            SystemSet::on_enter(self.on.clone()).with_system(start_loading::<Assets>.system()),
        )
        .add_system_set(
            SystemSet::on_update(self.on.clone())
                .with_system(check_loading_state::<State, Assets>.system()),
        );
    }
}

fn start_loading<Assets: AssetCollection>(mut commands: Commands, asset_server: Res<AssetServer>) {
    let mut handles = Assets::load(&asset_server);
    commands.insert_resource(LoadingAssetHandles {
        handles: handles.drain(..).map(|handle| handle.id).collect(),
        marker: PhantomData::<Assets>,
    })
}

fn check_loading_state<T: Component + Debug + Clone + Eq + Hash, Assets: AssetCollection>(
    mut commands: Commands,
    mut state: ResMut<State<T>>,
    next_state: Res<AssetLoaderNextState<T>>,
    asset_server: Res<AssetServer>,
    loading_asset_handles: Res<LoadingAssetHandles<Assets>>,
) {
    let load_state = asset_server.get_group_load_state(loading_asset_handles.handles.clone());
    if load_state == LoadState::Loaded {
        commands.insert_resource(Assets::create(&asset_server));
        state
            .set(next_state.next.clone())
            .expect("Failed to set next State");
    }
}
