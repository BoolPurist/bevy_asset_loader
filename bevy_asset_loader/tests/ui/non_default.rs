use bevy_asset_loader::*;
use bevy::prelude::*;

fn main() {}

#[derive(AssetCollection)]
struct Test {
    #[asset(path = "test.ogg")]
    test: Handle<AudioSource>,
    non_default: NoDefault
}

enum NoDefault {
    One,
    Two
}
