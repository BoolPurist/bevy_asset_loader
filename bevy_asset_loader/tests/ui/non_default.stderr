error[E0277]: the trait bound `NoDefault: Default` is not satisfied
   --> $DIR/non_default.rs:6:10
    |
6   | #[derive(AssetCollection)]
    |          ^^^^^^^^^^^^^^^ the trait `Default` is not implemented for `NoDefault`
    |
note: required by `std::default::Default::default`
    = note: this error originates in the derive macro `AssetCollection` (in Nightly builds, run with -Z macro-backtrace for more info)
